controller/konnect/reconciler_kongplugin.go

```diff
@@ -0,0 +1,278 @@
+					pluginBindingsToDelete = append(pluginBindingsToDelete, pb)
+				}
+			}
+
+			// iterate over all the KondService annotations
+			for _, pluginName := range pluginSlice {
+				if pluginName != kongPlugin.Name {
+					continue
+				}
+
+				pluginReferenceFound = true
+				// if the KongPluginBinding does not exist yet, create it
+				if len(pluginBindingsByServiceName[kongService.Name]) == 0 {
+					kongPluginBinding := configurationv1alpha1.KongPluginBinding{
+						ObjectMeta: metav1.ObjectMeta{
+							GenerateName: kongPlugin.Name + "-",
+							Namespace:    kongPlugin.Namespace,
+						},
+						Spec: configurationv1alpha1.KongPluginBindingSpec{
+							Targets: configurationv1alpha1.KongPluginBindingTargets{
+								ServiceReference: &configurationv1alpha1.TargetRefWithGroupKind{
+									Group: configurationv1alpha1.GroupVersion.Group,
+									Kind:  "KongService",
+									Name:  kongService.Name,
+								},
+							},
+							PluginReference: configurationv1alpha1.PluginRef{
+								Name: kongPlugin.Name,
+							},
+						},
+					}
+					k8sutils.SetOwnerForObject(&kongPluginBinding, &kongPlugin)
+					if err = r.client.Create(ctx, &kongPluginBinding); err != nil {
+						return ctrl.Result{}, err
+					}
+					log.Debug(logger, "Managed KongPluginBinding created", kongPluginBinding)
+				}
+				// add a finalizer to the KongService that means the associated kongPluginBindings need to be cleaned up
+				if controllerutil.AddFinalizer(&kongService, consts.CleanupPluginBindingFinalizer) {
+					if err = r.client.Update(ctx, &kongService); err != nil {
+						return ctrl.Result{}, err
+					}
+					log.Debug(logger, "KongService finalizer added", kongService)
+					return ctrl.Result{}, nil
+				}
+			}
+		}
+
+		// iterate over all the KongPluginBindings to be deleted and delete them.
+		for _, pb := range pluginBindingsToDelete {
+			if err = r.client.Delete(ctx, &pb); err != nil {
+				if k8serrors.IsNotFound(err) {
+					continue
+				}
+				return ctrl.Result{}, err
+			}
+			log.Info(logger, "KongPluginBinding deleted", pb)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// If some KongService is using the plugin, add a finalizer on the plugin.
+	// The KongPlugin cannot be deleted until all objects that reference it are
+	// deleted or do not reference it anymore.
+	if pluginReferenceFound {
+		if controllerutil.AddFinalizer(&kongPlugin, consts.PluginInUseFinalizer) {
+			if err = r.client.Update(ctx, &kongPlugin); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.Debug(logger, "KongPlugin finalizer added", kongPlugin)
+			return ctrl.Result{}, nil
+		}
+	} else {
+		if controllerutil.RemoveFinalizer(&kongPlugin, consts.PluginInUseFinalizer) {
+			if err = r.client.Update(ctx, &kongPlugin); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.Debug(logger, "KongPlugin finalizer removed", kongPlugin)
+			return ctrl.Result{}, nil
+		}
+	}
```

Comment:
How about extracting this to a separate method, e.g. `handleKongPluginFinalizer`?

Reply:
As we agreed, we're going to improve that in the upcoming PRs.